### PR TITLE
patchwork cleanups and enhacements

### PR DESCRIPTION
**Pull request by veruu**

### CODE REVIEW

File: sktm/patchwork.py [hunk 1]
```diff
@@ -418,6 +418,13 @@ def _get_mbox_url_sfx(self):
 
         return "mbox"
 
+    def _get_project_id(self, project_name):
+        """
+        Get project ID based on its name. Child classes need to implement
+        this method.
+        """
+        raise NotImplementedError
```
Comment (spbnick) [inline sktm/patchwork.py]: Can I ask you to write full documentation for this method? I think it's even more important to document than the implementations, as it defines the interface.

File: sktm/__init__.py [hunk 2]
```diff
@@ -124,7 +123,6 @@ def add_pw(self, baseurl, pname, lpatch=None, restapi=False, apikey=None,
                 baseurl, pname, int(lpatch) if lpatch else None, skip
             )
 
-            # FIXME Figure out the last patch first, then create the interface
```
Comment (spbnick) [inline sktm/__init__.py]: I didn't read this code now, but I think converting between project ID and name shouldn't be a job for a patchwork project class, but rather for a patchwork server (instance) class. The latter can have an interface to retrieve project objects for specific IDs and have an interface for conversion to and from a project name, i.e. the `_get_project_id()` function should be moved there. Then we will be able to deal with these FIXMEs properly.

That's not something we need to do in this PR, though.

File: sktm/executable.py [hunk 3]
```diff
@@ -106,8 +106,8 @@ def setup_parser():
     parser_patchwork.add_argument("repo", type=str, help="Base repo URL")
     parser_patchwork.add_argument("baseurl", type=str, help="Base URL")
     parser_patchwork.add_argument("project", type=str, help="Project name")
-    parser_patchwork.add_argument("--lastpatch", type=str, help="Last patch "
-                                  "(id for pw1; datetime for pw2)")
+    parser_patchwork.add_argument("--lastpatch", type=int,
+                                  help="Last patch ID")
```
Comment (spbnick) [inline sktm/executable.py]: Aaah, finally, thank you :relieved:

File: sktm/__init__.py [hunk 4]
```diff
@@ -92,9 +92,7 @@ def add_pw(self, baseurl, pname, lpatch=None, restapi=False, apikey=None,
         Args:
             baseurl:        Patchwork base URL.
             pname:          Patchwork project name.
-            lpatch:         Last processed patch. Patch ID, if adding an XML
-                            RPC-based interface. Patch timestamp, if adding a
-                            REST-based interface. Can be omitted to
+            lpatch:         Last processed patch. Patch ID. Can be omitted to
```
Comment (spbnick) [inline sktm/__init__.py]: Do you think saying `ID of the last processed patch.` here would read better?

File: sktm/patchwork.py [hunk 5]
```diff
@@ -902,8 +901,7 @@ def get_patch_by_id(self, pid):
             patch = self.rpc.patch_get(pid, self.fields)
 
         if patch is None or patch == {}:
-            logging.warning("Failed to get data for patch %d", pid)
-            patch = None
+            raise Exception('Can\'t get patch by id %d)'.format(pid))
```
Comment (spbnick) [inline sktm/patchwork.py]: Niiice :+1:

File: sktm/__init__.py (hunk 4 above)
Comment (veruu) [reply sktm/__init__.py]: Absolutely!

File: sktm/__init__.py (hunk 2 above)
Comment (veruu) [reply sktm/__init__.py]: Having a class to deal with a full Patchwork server is definitely a neat idea. Do you think you can submit an issue for that, so we don't forget about it while dealing with other things?

**Comment by veruu** [comment]
@spbnick ducs updated!

File: sktm/__init__.py (hunk 2 above)
Comment (spbnick) [reply sktm/__init__.py]: Aand, here it is: #140.

**Comment by spbnick** [comment]
Thank you, Veronika, and thank you for splitting the changes so nicely :)

**Pull request merged**